src/rgw/driver/sfs/types.h

```diff
@@ -38,52 +39,89 @@ namespace rgw::sal::sfs {
 
 struct UnknownObjectException : public std::exception { };
 
-
-struct Object {
-
+class Object {
+ public:
   struct Meta {
     size_t size;
     std::string etag;
     ceph::real_time mtime;
     ceph::real_time set_mtime;
     ceph::real_time delete_at;
-    std::map<std::string, bufferlist> attrs;
   };
 
   std::string name;
   std::string instance;
   uint version_id{0};
   UUIDPath path;
-  Meta meta;
   bool deleted;
 
-  Object(const std::string &_name)
-  : name(_name), path(UUIDPath::create()), deleted(false) { }
+ private:
+  Meta meta;
+  std::map<std::string, bufferlist> attrs;
 
-  Object(const std::string &_name, const uuid_d &_uuid, bool _deleted)
-  : name(_name), path(_uuid), deleted(_deleted) { }
+ protected:
+  Object(const std::string& _name, const uuid_d& _uuid);
 
-  Object(const rgw_obj_key &key)
-  : name(key.name), instance(key.instance), path(UUIDPath::create()), deleted(false) { }
+  Object(const rgw_obj_key& key)
+      : name(key.name),
+        instance(key.instance),
+        path(UUIDPath::create()),
+        deleted(false) {
+  }
+
+ public:
+  static Object* create_for_immediate_deletion(
+      const sqlite::DBOPObjectInfo& object);
+  static Object* create_for_query(const std::string& name, const uuid_d& uuid,
+                                  bool deleted, uint version_id);
+  static Object* create_for_testing(const std::string& name);
+  static Object* create_from_obj_key(const rgw_obj_key& key);
+  static Object* create_for_multipart(const std::string& name);
+
+  static Object* create_commit_delete_marker(const rgw_obj_key& key,
+                                             SFStore* store,
+                                             const std::string& bucket_id);
+  static Object* create_commit_new_object(const rgw_obj_key& key,
+                                          SFStore* store,
+                                          const std::string& bucket_id);
+
+  const Meta get_meta() const;
+  const Meta get_default_meta() const;
+  void update_meta(const Meta& update);
+
+  bool get_attr(const std::string& name, bufferlist& dest);
+  void set_attr(const std::string& name, bufferlist& value);
+  Attrs::size_type del_attr(const std::string& name);
+  Attrs get_attrs();
+  void update_attrs(const Attrs& update);
 
   std::filesystem::path get_storage_path() const;
 
-  void metadata_init(SFStore *store, const std::string & bucket_id,
-                     bool new_object, bool new_version);
-  void metadata_change_version_state(SFStore *store, ObjectState state);
-  void metadata_finish(SFStore *store);
+  // Add new object version. Sets version_id to the one created
+  void add_new_version(SFStore* store);
 
-  void metadata_flush_attrs(SFStore *store);
+  // Change obj version state.
```

Comment:
maybe turn these comments into doc strings? My IDE would benefit from that, at least :)